### PR TITLE
DATA-1887 & DATA-2012 Update collectors to use new proto responses and add tests for collectors

### PR DESCRIPTION
**Pull request by vijayvuyyuru**
Hey y'all!
The task is to change the collectors to use the new cleaner, prettier, and betterer proto responses. 
This PR is to change all collectors except sensor and board collectors as those 2 have some other changes that I am splitting out to a separate PR. 

**Testing**
Manually tested each collector here by running local RDK against app.viam.com. 
Screenshot_2023-09-29_at_12 34 40_PM-removebg-preview

### CODE REVIEW

File: components/motor/collectors.go
Comment (dmhilly) [inline components/motor/collectors.go:58]: This is moreso a question for @AaronCasas, but how did we decide which methods can have data capture configured? I see a lot of components have more methods than just the ones we have collectors registered for. For some of them, it doesn't make sense to have a collector since they don't actually return any data (for example `Stop`/`StopResponse`). But for motor for eg., `IsPowered` has a collector but `IsMoving` does not.

File: services/datamanager/builtin/capture_test.go
Comment (dmhilly) [inline services/datamanager/builtin/capture_test.go:271]: Surprised we don't need to change more tests - or do we not have tests on the output of each collector? I wonder if we should add that as part of this ticket (I think yes). 

Idea for a simple and concise way to test the collectors: we could call a particular method using gRPC and also via a collector and check that the response formats are the same.

File: components/movementsensor/collectors.go
Comment (dmhilly) [inline components/movementsensor/collectors.go:14]: Do we have a ticket for this TODO? (I know this was already here but figured I would ask 🙂)

File: components/movementsensor/collectors.go
Comment (dmhilly) [inline components/movementsensor/collectors.go:29]: What are these strings used for? Should they match the method names in the API?

File: components/movementsensor/collectors.go
Comment (dmhilly) [inline components/movementsensor/collectors.go:68]: Looks like we mixed up the linear velocity collector with the position collector (this function is called `newPositionCollector` but returns a linear velocity collector) 🙂

File: components/movementsensor/movementsensor.go
Comment (dmhilly) [inline components/movementsensor/movementsensor.go:28]: Oh I see this is where we use the String function - same question about the API names applies. I notice that for other components the names match the Go method names and not the API method names so there is a precedent set but I'm wondering why. Especially if this affects how users register collectors in their configs (not actually sure that's true, I'd need to dig into the rdk code some more!)

File: components/movementsensor/collectors.go
Comment (dmhilly) [inline components/movementsensor/collectors.go:52]: Just so I understand, you did not add new collectors for the movement sensor, but rather moved the collectors from `movementsensor.go` to here (`collectors.go`)?

And then ofc made the change to return the proto response format.

**Review by dmhilly — COMMENTED** [review]
Really nice work Vijay! This is a big change and I appreciate how diligent you've been about understanding each component type and desired output. Left some comments and questions; I didn't get to the powersensor yet but I'll review that this afternoon. I see in your PR description you mention a plan for manual testing, which is awesome. I do think we should at least attempt automatic testing of this as well so let me know if you'd like to discuss. Thanks for your hard work on this! 😄

File: components/arm/collectors.go
Comment (vijayvuyyuru) [reply components/arm/collectors.go]: I followed up with rdk and motion planning. Degrees is fine and what is expected.

File: components/gantry/collectors.go
Comment (AaronCasas) [inline components/gantry/collectors.go:83]: Suuuper nitty, but this looks possible inefficient compared do a direct calculation? Since for each value are basically writing twice instead of once (if we skipped the copy and just did `ret[i] = meters[i] * 1000`)

File: components/motor/collectors.go
Comment (AaronCasas) [reply components/motor/collectors.go:58]: No systematic way! When the original ones were added we added them for all methods that returned data. Since then, they've been added more adhoc

File: components/movementsensor/collectors.go
Comment (AaronCasas) [reply components/movementsensor/collectors.go:29]: I'd be in favor of them matching the API method names, but agreed that it's a separate change

**Review by AaronCasas — COMMENTED** [review]
Just a couple nits, then echoing @dmhilly on at least a simple test suite being useful!

File: services/datamanager/builtin/capture_test.go
Comment (vijayvuyyuru) [reply services/datamanager/builtin/capture_test.go:271]: @dmhilly @AaronCasas I talked to Alexa about the testing for collectors in general and this is the outcome:
Adding tests is definitely something we should do but is lower priority than project work. I'll make a ticket for testing and work on that separately maybe down the line but prioritizing whatever project work is there(which I know is how we got here LOL).

File: components/movementsensor/collectors.go
Comment (vijayvuyyuru) [reply components/movementsensor/collectors.go:14]: I will make this ticket and update the comment.

File: components/sensor/collector.go
Comment (vijayvuyyuru) [inline components/sensor/collector.go:45]: Spoke about this with @agreenb. We decide to bring sensor more in line with the API, we will no longer be filtering on sensor data, the collector will always get all values from the sensor

File: services/datamanager/builtin/capture_test.go
Comment (vijayvuyyuru) [reply services/datamanager/builtin/capture_test.go:271]: https://viam.atlassian.net/browse/DATA-2012

**Review by AaronCasas — COMMENTED** [review]
This LGTM, but we should hold off on merging until datasync and datamanegment are updated to write/read the new format, so holding off on approval until those tickets are ready

File: services/datamanager/builtin/capture_test.go
Comment (agreenb) [reply services/datamanager/builtin/capture_test.go:271]: We should make sure to test every changed collector, whether manually or via an automated test suite — initially assumed the former would be faster, but Devin's suggestion seems like the latter would likely be faster. Before merging these changes, it would actually be worth adding these tests to ensure we're only syncing data in the format that we expect.

Comment (agreenb) [reply services/datamanager/builtin/capture_test.go:271]: As discussed offline, whatever can be done within a day of focused time before merging these changes, which seems possible with the suggested testing methodology!

**Review by dmhilly — COMMENTED** [review]
Great work, just reviewed the rest of the code. Haven't looked at the tests yet and I know those aren't ready. Looking forward to taking a look once they are! 🙂

File: components/movementsensor/collectors.go
Comment (dmhilly) [reply components/movementsensor/collectors.go:29]: Filed [here](https://viam.atlassian.net/browse/DATA-2018)!

File: components/gantry/collectors.go
Comment (vijayvuyyuru) [reply components/gantry/collectors.go:84]: That is a great point, will fix when I push tests.

File: services/datamanager/builtin/capture_test.go
Comment (AaronCasas) [inline services/datamanager/builtin/capture_test.go:362]: If the goal here is to test the various component collectors produce correctly formatted data, is there a need to use the datamanager service as a pass through like this? What's the advantage vs directly building the collectors with the `newXCollector` constructors in each component/collectors.go file?

Comment (vijayvuyyuru) [reply services/datamanager/builtin/capture_test.go:362]: There isn't a particular need with respect to the goal. I did it here to be able to test all the collectors in the same place instead of having it per collector file and keeping all the configs in the same place. I can change it if thats better.

Comment (AaronCasas) [reply services/datamanager/builtin/capture_test.go:362]: Since the data manager is only being used as a sort of indirect passthrough to call start/stop on the underlying collector, I think it might obfuscate what we're actually testing more than it helps from being a bit more integration-y. So I would personally err towards putting it in some collector_test.go file in the component directories. I think this would allow you to not need the config files too, since you can just directly use the CollectorParams struct to build them. Very open to other opinions and arguments though (cc @agreenb @dmhilly)

Comment (dmhilly) [reply services/datamanager/builtin/capture_test.go:362]: Yeah I agree with you Aaron that that would be simpler. I may have put this idea in Vijay's head since I had thought the tests might work by running data capture and also making a grpc call to the same method and comparing the responses to make sure the format is the same. But I think comparing with the known grpc response message is doing pretty much the same thing and is simpler, so I'd be in favor of changing this to take out the datamanager layer, just constructing the component and passing it into the collector and testing that.

Comment (vijayvuyyuru) [reply services/datamanager/builtin/capture_test.go:362]: Makes sense, will do

File: components/arm/collectors_test.go
Comment (AaronCasas) [inline components/arm/collectors_test.go:22]: Nit/opinion: I think collectorConstructor would be a bit more accurate here. At first read, I assumed this was the capture function

File: components/gantry/collectors_test.go
Comment (AaronCasas) [inline components/gantry/collectors_test.go:71]: Can you make time.Second a const or var named `captureInterval` or `interval` or something? It's referenced a few places that need to match

File: components/arm/collectors_test.go
Comment (AaronCasas) [reply components/arm/collectors_test.go:22]: Looks like this is a copy of `data.CollectorConstructor` actually! I think you can just use that

File: components/powersensor/collectors_test.go
Comment (AaronCasas) [inline components/powersensor/collectors_test.go:117]: This is redefined in many places! Is there some single place it can be moved (testutils maybe?)?

File: components/gantry/collectors_test.go
Comment (AaronCasas) [reply components/gantry/collectors_test.go:71]: Likewise for other tests with this pattern

**Review by AaronCasas — DISMISSED** [review]
Just a few small things! Great job on this - I know it's a pretty annoying PR and appreicate you flying through it

File: components/powersensor/collectors_test.go
Comment (vijayvuyyuru) [reply components/powersensor/collectors_test.go:117]: Ill put it in test utils, I had a similar feeling about it.

File: components/encoder/collectors_test.go
Comment (AaronCasas) [inline components/encoder/collectors_test.go:48]: Maybe a dumb question because I haven't used it much - but I see we have all these very simple mocks for each of these. Is this something that that existing `inject` (or whatever it's called) package can already provide?

Comment (vijayvuyyuru) [reply components/encoder/collectors_test.go:48]: Importing the inject package creates a circular import loop as the inject package imports the component.

File: components/arm/collectors_test.go
Comment (dmhilly) [inline components/arm/collectors_test.go:41]: Minor but since the params are the same between the two tests, you can remove this field from the test struct and just construct them in the test loop below

File: components/gantry/collectors_test.go
Comment (dmhilly) [inline components/gantry/collectors_test.go:34]: Same feedback here and in a couple other places about the params!

File: components/motor/collectors_test.go
Comment (dmhilly) [inline components/motor/collectors_test.go:73]: I wonder if you could factor this out into a testutils function somewhere that takes a collector function as an argument. It seems to be the same across the different collectors packages

Then here you could  just call something like `testutils.RunCollectorTest(t, newMotor, ...)`

I don't want to create much more work for you though since this has been a big ticket :) So just an idea.

File: testutils/file_utils.go
Comment (dmhilly) [inline testutils/file_utils.go:33]: Nice 🔥

**Review by dmhilly — COMMENTED** [review]
Really nice work Vijay! Just reviewed the tests and they look great. Left a couple comments but looks like this is very close to done! 🤞

File: components/motor/collectors_test.go
Comment (vijayvuyyuru) [reply components/motor/collectors_test.go:87]: I assumed that because a collector needed a component name that I should leave it in each component. In all honesty I didn't really check if this was the case. Ill test if I need them and if not, remove them.

File: testutils/file_utils.go
Comment (agreenb) [reply testutils/file_utils.go:33]: +1, super clean for testing!

File: testutils/file_utils.go
Comment (agreenb) [inline testutils/file_utils.go:55]: [supernit] just in case the path ends up being validated later, can make this a valid path, e.g. `/mock/dir`

File: testutils/proto_utils.go
Comment (agreenb) [inline testutils/proto_utils.go:7]: I believe you meant to use an an `*anypb.Any` here, whereas `any` is a lesser known alias to `interface{}`. Since StructToStructPb takes in an `interface{}` and `AsMap()` returns a map[string]interface{}, would explicitly use interface{} here since it's much more conventional (and thus readable) in Go — or just explicitly go with anypb.Any.

Then just nit to update the comment above.

File: components/arm/collectors_test.go
Comment (agreenb) [inline components/arm/collectors_test.go:38]: Nice, really clean with the new IgnoreOmitEmpty util you added!

File: components/encoder/collectors_test.go
Comment (agreenb) [reply components/encoder/collectors_test.go:46]: We discussed with @dmhilly about this offline as a clean way to get around importing the injected arm and having the circular dependency. The only thing I worry about is if the Go method signature changes, especially the internal Go struct return types — but if it did, then the collector code itself shouldn't compile in the first place.

Another way to included the `inject` package would be to make this test file in a different package and export the `newXCollector` constructors. We already exported `RegisterCollector` in the `data` package, and I don't think there's much harm — if another package tries to reregister on the same method, we correctly return an error (or more specifically, panic).

File: components/movementsensor/collectors_test.go
Comment (agreenb) [inline components/movementsensor/collectors_test.go:148]: Nice helpers all throughout the code

File: components/sensor/collector.go
Comment (agreenb) [reply components/sensor/collector.go:45]: Note that before merging this PR, will post in #eng-breaking-changes and #solutions-engineering.

**Review by agreenb — COMMENTED** [review]
This looks great! Echoing the others: you've been extremely diligent, thorough, and lightning fast in all of your implementation. Fantastic work.

Just a few last comments/questions, and then I'd ask to try collecting all of these into app.viam.com again to ensure it's all being uploaded as expected.

File: testutils/proto_utils.go
Comment (vijayvuyyuru) [reply testutils/proto_utils.go:7]: I'll update it to be an interface. I meant to just take in any data without converting to an anypb, will update and push.

File: components/encoder/collectors_test.go
Comment (AaronCasas) [reply components/encoder/collectors_test.go:46]: > Another way to included the inject package would be to make this test file in a different package and export the newXCollector constructors. We already exported RegisterCollector in the data package, and I don't think there's much harm — if another package tries to reregister on the same method, we correctly return an error (or more specifically, panic).

To clarify - the suggestion is to move collectors.go and collectors_test.go to a new package (or just the latter)? I think moving both would make sense, and (at least at first glance) I don't see anything that would break

Comment (agreenb) [reply components/encoder/collectors_test.go:46]: Just the latter, since I believe moving both would create a circular dependency where the collector brings in the component but the component needs to depend on the collector.

Comment (agreenb) [reply components/encoder/collectors_test.go:46]: Chatted offline and Vijay's making these changes — thank you! This'll keep the test code robust to any method signature changes on the component.

**Comment by vijayvuyyuru** [comment]
Tested manually on app against latest changes. All data looks correct and is displaying as expected in my org.

File: components/arm/collectors_test.go
Comment (agreenb) [inline components/arm/collectors_test.go:88]: So fresh, so clean, so robust. Thanks for the super quick iteration on this! 🚀

File: components/encoder/collectors.go
Comment (agreenb) [inline components/encoder/collectors.go:26]: position -> ticks count

File: components/movementsensor/collectors.go
Comment (agreenb) [inline components/movementsensor/collectors.go:14]: Can remove this TODO! 🪨 ⭐

**Review by agreenb — APPROVED** [review]
Thanks for retesting all of the collectors manually to ensure that the data uploaded to app was as expected. Fantastic work on tackling a big project with many moving parts, iterating quickly based on feedback, and writing clean code and tests.

LGTM! 🚀  Ask that you hold off on merging this and the board PR together after we've posted in the Slack channels.

File: components/board/collector.go
Comment (agreenb) [inline components/board/collector.go:40]: Since this is repeated 3 times, extract into a const? Also, if you can add "Must supply `reader_name` in additional params for analog collector" to make it clear what the user needs to update

Comment (agreenb) [reply components/board/collector.go:40]: It'd be good to also catch this at reconfiguration time. We can start validating [here](https://github.com/viamrobotics/rdk/blob/b34d7db3573c3bb9e0d7dbeddaa1c16407bf1b05/services/datamanager/builtin/builtin.go#L205-L216) by mapping `MethodMetadata` to expected params (locally on a fake board, you can print the method metadata to get what it should look like). Since this is the only collector that's using additional params right now, this would only contain `reader_name`.

File: components/board/collector.go
Comment (agreenb) [inline components/board/collector.go:76]: I think it makes sense for the above AnalogCollector to be `reader_name` but here to be `pin_name`.

File: components/board/collector.go
Comment (agreenb) [inline components/board/collector.go:32]: (fyi) I'll do this closer to the RDK release cut to coordinate with app, but we'll want to rename the above to the actual API methods: ReadAnalogReader and GetGPIO

**Review by agreenb — COMMENTED** [review]
Just one request with additional params validation at reconfiguration time, and one tiny nit. Really great work!

File: components/board/collector.go
Comment (vijayvuyyuru) [reply components/board/collector.go:40]: But even for the board collectors, one is `reader_name` the other is `pin_name`. So would I just need to enforce that additional params should contain one? Do I also need to enforce this on every collector or would it special case to only validate if it is a board component?

Comment (agreenb) [reply components/board/collector.go:40]: MethodMetadata contains the component + method, so the board + reader would map to `reader_name` and board + gpio would map to `pin_name`. If the method metadata exists in the dict, you check that it contains the additional params listed. Let me know if that makes sense!

File: services/datamanager/builtin/builtin.go
Comment (agreenb) [inline services/datamanager/builtin/builtin.go:256]: Rather than specifying board, you can use `md.MethodMetadata.API.String()` (and then remove the 'only check if board' comment above). When this is updated, could you retest manually on the board collectors into app? Thanks Vijay!

**Review by agreenb — APPROVED** [review]
Thanks for the additional thorough testing of board with/without correct additional params and other components since your latest changes! All LGTM, fantastic superb work 🚀  Will reach back out to you to merge this once the sync PR is ready.

**Review by dmhilly — APPROVED** [review]
Apologies for not getting to this before the weekend, but this LGTM. Excellent work Vijay, thanks so much for your contributions! 😄

**Comment by dmhilly** [comment]
Added @jeremyrhyde per @tessavitabile's request to make sure the SLAM team knows what changes are happening here that might impact the replay IMU and replay movement sensor.

**Review by jeremyrhyde — APPROVED** [review]
LGTM, very clean

**Review by dmhilly — CHANGES_REQUESTED** [review]
Leaving a comment here so we don't forget - we need to change the code to not structpb-ify the structpb.Values that are in the GetReadingsResponse

Edit: done!

**Review by dmhilly — APPROVED** [review]
LGTM! Looks like there are a couple failures still; lmk if you'd like me to help troubleshoot.

**Pull request merged**
